File: ios/ExpoMapboxNavigationModule.swift
import ExpoModulesCore
import CoreLocation

public class ExpoMapboxNavigationModule: Module {

  private lazy var offlineTileManager = ExpoMapboxNavigationOfflineManager()

  public func definition() -> ModuleDefinition {
    Name("ExpoMapboxNavigation")

    // Offline tile pre-download API (MOB-383). Module-level (not view) because downloads run on the
    // prep screen before any MapboxNavigationView is mounted.
    Events("onOfflineProgress", "onOfflineComplete", "onOfflineError")

    AsyncFunction("downloadOfflineRegion") { (options: [String: Any], promise: Promise) in
      guard
        let regionId = options["regionId"] as? String,
        let geometry = options["geometry"] as? [String: Any],
        let coordinates = geometry["coordinates"] as? [[[Double]]],
        let styleURL = options["styleURL"] as? String
      else {
        promise.reject("ERR_OFFLINE_ARGS", "Missing or invalid offline region options")
        return
      }
      // Clamp to a valid tile zoom range (0...22) before converting — UInt8(...) traps on a
      // negative or >255 value — and normalize order so minZoom <= maxZoom (a reversed range traps).
      let rawMinZoom = min(max((options["minZoom"] as? Int) ?? 7, 0), 22)
      let rawMaxZoom = min(max((options["maxZoom"] as? Int) ?? 15, 0), 22)
      let minZoom = UInt8(min(rawMinZoom, rawMaxZoom))
      let maxZoom = UInt8(max(rawMinZoom, rawMaxZoom))

      // Mapbox offline operations deliver their progress/completion callbacks on the initiating
      // thread's run loop; the Expo AsyncFunction closure runs on a transient background queue that
      // has none, so the callbacks would never fire. Run on the main queue (mirrors the Android side,
      // which dispatches to Dispatchers.Main).
      DispatchQueue.main.async {
        self.offlineTileManager.downloadRegion(
          regionId: regionId,
          geometryCoordinates: coordinates,
          styleURLString: styleURL,
          minZoom: minZoom,
          maxZoom: maxZoom,
          progress: { [weak self] (rid, stage, completed, required) in
            self?.sendEvent("onOfflineProgress", [
              "regionId": rid,
              "stage": stage,
              "completedResourceCount": Int(completed),
              "requiredResourceCount": Int(required),
            ])
          },
          completion: { [weak self] error in
            if let error {
              self?.sendEvent("onOfflineError", ["regionId": regionId, "message": error.localizedDescription])
              promise.reject("ERR_OFFLINE_DOWNLOAD", error.localizedDescription)
            } else {
              self?.sendEvent("onOfflineComplete", ["regionId": regionId])
              promise.resolve(["regionId": regionId])
            }
          }
        )
      }
    }

    AsyncFunction("removeOfflineRegion") { (regionId: String, styleURL: String?, promise: Promise) in
      DispatchQueue.main.async {
        self.offlineTileManager.removeRegion(regionId: regionId, styleURLString: styleURL) {
          promise.resolve(nil)
        }
      }
    }

    AsyncFunction("getOfflineRegions") { (promise: Promise) in
      DispatchQueue.main.async {
        self.offlineTileManager.listRegions { regions in
          promise.resolve(regions)
        }
      }
    }

    View(ExpoMapboxNavigationView.self) {
      Events("onRouteProgressChanged", "onCancelNavigation", "onWaypointArrival", "onFinalDestinationArrival", "onRouteChanged", "onUserOffRoute", "onRoutesLoaded", "onRouteFailedToLoad", "onLocationChange", "onMuteChange")

      Prop("coordinates") { (view: ExpoMapboxNavigationView, coordinates: Array<Dictionary<String, Any>>) in
         var points: Array<CLLocationCoordinate2D> = []
         for coordinate in coordinates {
            let longValue = coordinate["longitude"]
            let latValue = coordinate["latitude"]
            if let long = longValue as? Double, let lat = latValue as? Double {
                points.append(CLLocationCoordinate2D(latitude: lat, longitude: long))
            }
          }
          view.controller.setCoordinates(coordinates: points)
      }

      Prop("vehicleMaxHeight") { (view: ExpoMapboxNavigationView, maxHeight: Double?) in
          view.controller.setVehicleMaxHeight(maxHeight: maxHeight)
      }

      Prop("vehicleMaxWidth") { (view: ExpoMapboxNavigationView, maxWidth: Double?) in
          view.controller.setVehicleMaxWidth(maxWidth: maxWidth)
      }

      Prop("vehicleMaxWeight") { (view: ExpoMapboxNavigationView, maxWeight: Double?) in
          view.controller.setVehicleMaxWeight(maxWeight: maxWeight)
      }

      Prop("allowsArrivingOnOppositeSide") { (view: ExpoMapboxNavigationView, allows: Bool?) in
          view.controller.setAllowsArrivingOnOppositeSide(allows: allows)
      }

      Prop("showsEndOfRouteFeedback") { (view: ExpoMapboxNavigationView, shows: Bool?) in
          view.controller.setShowsEndOfRouteFeedback(shows: shows)
      }

      Prop("hideTripProgress") { (view: ExpoMapboxNavigationView, hide: Bool?) in
          view.controller.setHideTripProgress(hide: hide)
      }

      Prop("locale") { (view: ExpoMapboxNavigationView, locale: String?) in
          view.controller.setLocale(locale: locale)
      }

      Prop("useRouteMatchingApi"){ (view: ExpoMapboxNavigationView, useRouteMatchingApi: Bool?) in
          view.controller.setIsUsingRouteMatchingApi(useRouteMatchingApi: useRouteMatchingApi)
      }

      Prop("waypointIndices"){ (view: ExpoMapboxNavigationView, indices: Array<Int>?) in
          view.controller.setWaypointIndices(waypointIndices: indices)
      }

      Prop("routeProfile"){ (view: ExpoMapboxNavigationView, profile: String?) in
          view.controller.setRouteProfile(profile: profile)
      }

      Prop("routeExcludeList"){ (view: ExpoMapboxNavigationView, excludeList: Array<String>?) in
          view.controller.setRouteExcludeList(excludeList: excludeList)
      }

      Prop("mapStyle"){ (view: ExpoMapboxNavigationView, style: String?) in
          view.controller.setMapStyle(style: style)
      }

      Prop("mute"){ (view: ExpoMapboxNavigationView, isMuted: Bool?) in
          view.controller.setIsMuted(isMuted: isMuted)
      }

      Prop("initialLocation") { (view: ExpoMapboxNavigationView, location: Dictionary<String, Any>?) in
        if(location != nil){
          let longValue = location!["longitude"]
          let latValue = location!["latitude"]
          let zoomValue = location!["zoom"]
          if let long = longValue as? Double, let lat = latValue as? Double, let zoom = zoomValue as? Double? {
              view.controller.setInitialLocation(location: CLLocationCoordinate2D(latitude: lat, longitude: long), zoom: zoom)
          }
        }
      }

      Prop("customRasterSourceUrl") { (view: ExpoMapboxNavigationView, url: String?) in
        view.controller.setCustomRasterSourceUrl(url: url)
      }

      Prop("placeCustomRasterLayerAbove") { (view: ExpoMapboxNavigationView, layerId: String?) in
        view.controller.setPlaceCustomRasterLayerAbove(layerId: layerId)
      }

      Prop("disableAlternativeRoutes") { (view: ExpoMapboxNavigationView, disableAlternativeRoutes: Bool?) in
        view.controller.setDisableAlternativeRoutes(disableAlternativeRoutes: disableAlternativeRoutes)
      }

      Prop("followingZoom") { (view: ExpoMapboxNavigationView, followingZoom: Double?) in
        view.controller.setFollowingZoom(followingZoom: followingZoom)
      }

      Prop("uiStyle") { (view: ExpoMapboxNavigationView, uiStyle: String?) in
        view.controller.setUIStyle(style: uiStyle)
      }

      AsyncFunction("recenterMap") { (view: ExpoMapboxNavigationView) in
        view.controller.recenterMap()
      }
    }
  }
}

File: android/src/main/java/expo/modules/mapboxnavigation/ExpoMapboxNavigationModule.kt
package expo.modules.mapboxnavigation

import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import com.mapbox.common.TileStore
import com.mapbox.geojson.Point
import com.mapbox.navigation.base.options.NavigationOptions
import com.mapbox.navigation.base.options.RoutingTilesOptions
import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class ExpoMapboxNavigationModule : Module() {
  private val activity
    get() = requireNotNull(appContext.activityProvider?.currentActivity)

  // A single shared TileStore backs both the Nav SDK's routing tiles (via RoutingTilesOptions
  // below) and the Maps SDK display tiles / offline downloads, so offline routing and the base map
  // read from the same on-device store (MOB-383).
  private val sharedTileStore: TileStore by lazy { TileStore.create() }
  private val offlineTileManager by lazy { OfflineTileManager(sharedTileStore) }

  @com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
  override fun definition() = ModuleDefinition {
    Name("ExpoMapboxNavigation")

    OnActivityEntersForeground {
      (activity as LifecycleOwner).lifecycleScope.launch(Dispatchers.Main) {
        if (!MapboxNavigationApp.isSetup()) {
          MapboxNavigationApp.setup {
            NavigationOptions.Builder(activity.applicationContext)
              .routingTilesOptions(
                RoutingTilesOptions.Builder().tileStore(sharedTileStore).build()
              )
              .build()
          }
        }
        MapboxNavigationApp.attach(activity as LifecycleOwner)
      }
    }

    // Offline tile pre-download API (MOB-383). Module-level (not view) because downloads run on the
    // prep screen before any ExpoMapboxNavigationView is mounted.
    Events("onOfflineProgress", "onOfflineComplete", "onOfflineError")

    AsyncFunction("downloadOfflineRegion") { options: Map<String, Any?>, promise: Promise ->
      val regionId = options["regionId"] as? String
      val geometry = options["geometry"] as? Map<*, *>
      @Suppress("UNCHECKED_CAST")
      val coordinates = geometry?.get("coordinates") as? List<List<List<Double>>>
      val styleURL = options["styleURL"] as? String
      if (regionId == null || coordinates == null || styleURL == null) {
        promise.reject("ERR_OFFLINE_ARGS", "Missing or invalid offline region options", null)
        return@AsyncFunction
      }
      val minZoom = (options["minZoom"] as? Number)?.toInt() ?: 7
      val maxZoom = (options["maxZoom"] as? Number)?.toInt() ?: 15

      (activity as LifecycleOwner).lifecycleScope.launch(Dispatchers.Main) {
        offlineTileManager.downloadRegion(
          regionId,
          coordinates,
          styleURL,
          minZoom,
          maxZoom,
          { rid, stage, completed, required ->
            sendEvent(
              "onOfflineProgress",
              mapOf(
                "regionId" to rid,
                "stage" to stage,
                "completedResourceCount" to completed,
                "requiredResourceCount" to required
              )
            )
          },
          { error ->
            if (error != null) {
              sendEvent("onOfflineError", mapOf("regionId" to regionId, "message" to error))
              promise.reject("ERR_OFFLINE_DOWNLOAD", error, null)
            } else {
              sendEvent("onOfflineComplete", mapOf("regionId" to regionId))
              promise.resolve(mapOf("regionId" to regionId))
            }
          }
        )
      }
    }

    AsyncFunction("removeOfflineRegion") { regionId: String, styleURL: String?, promise: Promise ->
      offlineTileManager.removeRegion(regionId, styleURL) { promise.resolve(null) }
    }

    AsyncFunction("getOfflineRegions") { promise: Promise ->
      offlineTileManager.listRegions { regions -> promise.resolve(regions) }
    }

    View(ExpoMapboxNavigationView::class) {
      Events(
              "onRouteProgressChanged",
              "onCancelNavigation",
              "onWaypointArrival",
              "onFinalDestinationArrival",
              "onRouteChanged",
              "onUserOffRoute",
              "onRoutesLoaded",
              "onRouteFailedToLoad",
              "onLocationChange",
              "onMuteChange"
      )

      Prop("coordinates") { view: ExpoMapboxNavigationView, coordinates: List<Map<String, Any>> ->
        val points = mutableListOf<Point>()
        for (coordinate in coordinates) {
          val longValue = coordinate.get("longitude")
          val latValue = coordinate.get("latitude")
          if (longValue is Double && latValue is Double) {
            points.add(Point.fromLngLat(longValue, latValue))
          }
        }
        view.setCoordinates(points)
      }

      Prop("vehicleMaxHeight") { view: ExpoMapboxNavigationView, maxHeight: Double? ->
        view.setVehicleMaxHeight(maxHeight)
      }

      Prop("vehicleMaxWidth") { view: ExpoMapboxNavigationView, maxWidth: Double? ->
        view.setVehicleMaxWidth(maxWidth)
      }

      Prop("vehicleMaxWeight") { view: ExpoMapboxNavigationView, maxWeight: Double? ->
        view.setVehicleMaxWeight(maxWeight)
      }

      Prop("allowsArrivingOnOppositeSide") { view: ExpoMapboxNavigationView, allows: Boolean? ->
        view.setAllowsArrivingOnOppositeSide(allows)
      }

      Prop("showsEndOfRouteFeedback") { view: ExpoMapboxNavigationView, shows: Boolean? ->
        view.setShowsEndOfRouteFeedback(shows)
      }

      Prop("hideTripProgress") { view: ExpoMapboxNavigationView, hide: Boolean? ->
        view.setHideTripProgress(hide)
      }

      Prop("waypointIndices") { view: ExpoMapboxNavigationView, indices: List<Int>? ->
        view.setWaypointIndices(indices)
      }

      Prop("locale") { view: ExpoMapboxNavigationView, localeStr: String? ->
        view.setLocale(localeStr)
      }

      Prop("useRouteMatchingApi") { view: ExpoMapboxNavigationView, useRouteMatchingApi: Boolean? ->
        view.setIsUsingRouteMatchingApi(useRouteMatchingApi)
      }

      Prop("routeProfile") { view: ExpoMapboxNavigationView, profile: String? ->
        view.setRouteProfile(profile)
      }

      Prop("routeExcludeList") { view: ExpoMapboxNavigationView, excludeList: List<String>? ->
        view.setRouteExcludeList(excludeList)
      }

      Prop("mapStyle") { view: ExpoMapboxNavigationView, style: String? -> view.setMapStyle(style) }

      Prop("mute") { view: ExpoMapboxNavigationView, isMuted: Boolean? -> view.setIsMuted(isMuted) }

      Prop("initialLocation") { view: ExpoMapboxNavigationView, initialLocation: Map<String, Any>?
        ->
        val longValue = initialLocation?.get("longitude")
        val latValue = initialLocation?.get("latitude")
        val zoomValue = initialLocation?.get("zoom")

        if (longValue is Double && latValue is Double && zoomValue is Double?) {
          view.setInitialLocation(Point.fromLngLat(longValue, latValue), zoomValue)
        }
      }

      Prop("customRasterSourceUrl") { view: ExpoMapboxNavigationView, url: String? ->
        view.setCustomRasterSourceUrl(url)
      }

      Prop("placeCustomRasterLayerAbove") { view: ExpoMapboxNavigationView, layerId: String? ->
        view.setPlaceCustomRasterLayerAbove(layerId)
      }

      Prop("disableAlternativeRoutes") {
              view: ExpoMapboxNavigationView,
              disableAlternativeRoutes: Boolean? ->
        view.setDisableAlternativeRoutes(disableAlternativeRoutes)
      }

      Prop("followingZoom") { view: ExpoMapboxNavigationView, followingZoom: Double? ->
        view.setFollowingZoom(followingZoom)
      }

      AsyncFunction("recenterMap") { view: ExpoMapboxNavigationView -> view.recenterMap() }
    }
  }
}
